aten/src/ATen/core/ivalue_inl.h

```diff
@@ -578,6 +578,22 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
     return c10::make_intrusive<Future>(std::move(type), devices_);
   }
 
+  ~Future() {
+    // If there are no more references to this future (and thus the refcount
+    // drops to zero and the destructor is called) while some callbacks are
```

Comment:
When do we expect this to happen? IIRC, RPC will hold its futures alive, and chained futures will be kept alive within the callback. Does this mean the only way this could happen would be the following? But isn't this an user code error?

```python
x = Future().then()
x.wait()
```

Reply:
I was thinking of cases like this:
```
def foo():
    f = Future()
    if some_condition:
        defer_mark_complete(f)
    return f
```
which is a bug because if `some_condition` isn't satisfied no one will ever mark `f` complete. I suspect this can become hard to spot once the code becomes complex enough.